src/main.rs

```diff
@@ -179,60 +171,49 @@ fn main() -> Result<()> {
                 Err(e) => {
-                    fatal_error(format!("Cannot create sigstore verifier: {:?}", e));
+                    fatal_error(e.to_string());
                     unreachable!()
                 }
-            };
+                Ok(v) => Some(v),
+            }
+        } else {
+            None
+        };
+
+        for (name, policy) in policies.iter_mut() {
+            debug!(policy = name.as_str(), "download");
 
             let mut verified_manifest_digest: Option<String> = None;
 
-            if verify_enabled {
+            if let Some(ver) = verifier.as_mut() {
                 info!(
                     policy = name.as_str(),
                     "verifying policy authenticity and integrity using sigstore"
                 );
-
-                // verify policy prior to pulling for all keys, and keep the
-                // verified manifest digest of last iteration, even if all are
-                // the same:
-                for key_value in verification_settings
-                    .as_ref()
-                    .unwrap()
-                    .verification_keys
-                    .values()
-                {
-                    verified_manifest_digest = Some(
-                        verifier
-                            .verify(
-                                &policy.url,
-                                docker_config.clone(),
-                                verification_settings
-                                    .as_ref()
-                                    .unwrap()
-                                    .verification_annotations
-                                    .clone(),
-                                key_value,
-                            )
-                            .await
-                            .map_err(|e| {
-                                fatal_error(format!(
-                                    "Policy '{}' cannot be verified: {:?}",
-                                    name, e
-                                ))
-                            })
-                            .unwrap(),
-                    );
-                }
+                let verification_config = match cli::verification_settings(&matches) {
+                    Ok(config) => config,
+                    Err(e) => {
+                        fatal_error(format!(
+                            "Cannot create sigstore verification config: {:?}",
+                            e
+                        ));
+                        unreachable!()
+                    }
+                };
+                verified_manifest_digest = Some(
+                    ver.verify(&policy.url, docker_config.clone(), verification_config)
+                        .await
+                        .map_err(|e| {
+                            fatal_error(format!("Policy '{}' cannot be verified: {:?}", name, e))
+                        })
+                        .unwrap(),
```

Comment:
In case of more than one offender this is only reporting the first one and exiting, right? I think it would be nice to report all offenders at once.

Reply:
No, it prints all the constraints from the verification_config that are not satisfied. E.g:

https://github.com/kubewarden/policy-fetcher/blob/6fe85ba4f7a64e60171aead5511cea34d78d261d/src/verify/mod.rs#L526-L539

Reply:
yes, but it will exit (fatal_error) as soon as one policy fails without evaluating the others policies pending in the loop